File: packages/sdk-typescript/src/trace.ts
import { ENDPOINTS, LOG_PREFIX } from './config.js';
import { post } from './http.js';
import type {
  ErrorOptions,
  EventBase,
  EventPayload,
  EventType,
  LLMResponseOptions,
  LLMThinkingOptions,
  MessageItem,
  EmbeddingRequestOptions,
  EmbeddingResponseOptions,
  ToolCallPair,
  ToolCallRequestOptions,
  ToolCallResponseOptions,
  ToolResultOptions,
  UserMessageOptions,
} from './types.js';

function uuid(): string {
  if (typeof crypto !== 'undefined' && crypto.randomUUID) {
    return crypto.randomUUID();
  }
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
    const r = (Math.random() * 16) | 0;
    return (c === 'x' ? r : (r & 0x3) | 0x8).toString(16);
  });
}

export class WhyOpsTrace {
  constructor(
    private readonly traceId: string,
    private readonly agentName: string,
    private readonly apiKey: string,
    private readonly analyseBaseUrl: string,
    private readonly onInit: () => Promise<void>,
  ) {}

  // ─── user_message ────────────────────────────────────────────────────────

  async userMessage(messages: MessageItem[], options: UserMessageOptions = {}): Promise<void> {
    const meta: Record<string, unknown> | undefined = options.metadata
      ? { ...options.metadata }
      : undefined;
    await this.send(this.buildPayload('user_message', { messages }, meta, options));
  }

  // ─── llm_response ────────────────────────────────────────────────────────

  async llmResponse(
    model: string,
    provider: string,
    content: string | null | undefined,
    options: LLMResponseOptions = {},
  ): Promise<void> {
    const meta: Record<string, unknown> = { model, provider };
    if (options.usage) meta['usage'] = options.usage;
    if (options.latencyMs !== undefined) meta['latencyMs'] = options.latencyMs;

    const eventContent: Record<string, unknown> = {};
    if (content != null) eventContent['content'] = content;
    if (options.toolCalls?.length) eventContent['toolCalls'] = options.toolCalls;
    if (options.finishReason) eventContent['finishReason'] = options.finishReason;

    await this.send(this.buildPayload('llm_response', eventContent, meta, options));
  }

  // ─── llm_thinking ────────────────────────────────────────────────────────

  async llmThinking(thinking: string, options: LLMThinkingOptions = {}): Promise<void> {
    const c: Record<string, unknown> = { type: 'thinking', thinking };
    if (options.signature) c['signature'] = options.signature;
    await this.send(this.buildPayload('llm_thinking', c, undefined, options));
  }

  // ─── embedding_request ───────────────────────────────────────────────────

  async embeddingRequest(inputs: string[], options: EmbeddingRequestOptions = {}): Promise<void> {
    await this.send(this.buildPayload('embedding_request', { input: inputs }, undefined, options));
  }

  // ─── embedding_response ──────────────────────────────────────────────────

  async embeddingResponse(
    model: string,
    provider: string,
    embeddingCount: number,
    firstEmbeddingDimensions: number,
    options: EmbeddingResponseOptions = {},
  ): Promise<void> {
    const meta: Record<string, unknown> = { model, provider };
    if (options.totalTokens !== undefined) meta['usage'] = { totalTokens: options.totalTokens };
    if (options.latencyMs !== undefined) meta['latencyMs'] = options.latencyMs;

    await this.send(this.buildPayload(
      'embedding_response',
      { object: 'list', embeddingCount, firstEmbeddingDimensions, encodingFormat: 'float' },
      meta,
      options,
    ));
  }

  // ─── tool_call_request ───────────────────────────────────────────────────

  /**
   * Emit a tool_call_request event.
   * Returns the spanId — pass it to `toolCallResponse` to pair them.
   */
  async toolCallRequest(
    tool: string,
    toolCalls: ToolCallPair[],
    options: ToolCallRequestOptions = {},
  ): Promise<string> {
    const spanId = options.spanId ?? uuid();
    const c: Record<string, unknown> = { toolCalls };
    if (options.requestedAt) c['requestedAt'] = options.requestedAt;
    const meta: Record<string, unknown> = { tool };
    if (options.latencyMs !== undefined) meta['latencyMs'] = options.latencyMs;

    await this.send(this.buildPayload('tool_call_request', c, meta, { ...options, spanId }));
    return spanId;
  }

  // ─── tool_call_response ──────────────────────────────────────────────────

  async toolCallResponse(
    tool: string,
    spanId: string,
    toolCalls: ToolCallPair[],
    toolResults: Record<string, unknown>,
    options: ToolCallResponseOptions = {},
  ): Promise<void> {
    const c: Record<string, unknown> = { toolCalls, toolResults };
    if (options.respondedAt) c['respondedAt'] = options.respondedAt;
    const meta: Record<string, unknown> = { tool };
    if (options.latencyMs !== undefined) meta['latencyMs'] = options.latencyMs;

    await this.send(this.buildPayload('tool_call_response', c, meta, { ...options, spanId }));
  }

  // ─── tool_result ─────────────────────────────────────────────────────────

  async toolResult(
    toolName: string,
    output: Record<string, unknown>,
    options: ToolResultOptions = {},
  ): Promise<void> {
    await this.send(this.buildPayload('tool_result', { toolName, output }, undefined, options));
  }

  // ─── error ───────────────────────────────────────────────────────────────

  async error(message: string, options: ErrorOptions = {}): Promise<void> {
    const c: Record<string, unknown> = { message };
    if (options.status !== undefined) c['status'] = options.status;
    if (options.stack) c['stack'] = options.stack;
    await this.send(this.buildPayload('error', c, undefined, options));
  }

  // ─── Internal ────────────────────────────────────────────────────────────

  private buildPayload(
    eventType: EventType,
    content: unknown,
    metadata: Record<string, unknown> | undefined,
    opts: EventBase,
  ): EventPayload {
    const payload: EventPayload = {
      eventType,
      traceId: this.traceId,
      agentName: this.agentName,
      content,
    };
    if (opts.spanId) payload.spanId = opts.spanId;
    if (opts.stepId) payload.stepId = opts.stepId;
    if (opts.parentStepId) payload.parentStepId = opts.parentStepId;
    if (opts.timestamp) payload.timestamp = opts.timestamp;
    if (opts.idempotencyKey) payload.idempotencyKey = opts.idempotencyKey;
    if (metadata) payload.metadata = metadata;
    return payload;
  }

  private async send(payload: EventPayload): Promise<void> {
    await this.onInit();
    try {
      const res = await post(
        `${this.analyseBaseUrl}${ENDPOINTS.eventsIngest}`,
        payload,
        { Authorization: `Bearer ${this.apiKey}` },
      );
      if (!res.ok) {
        console.error(`${LOG_PREFIX} event send failed: HTTP ${res.status}`, payload.eventType);
      }
    } catch (err) {
      console.error(`${LOG_PREFIX} event send error:`, err);
    }
  }
}

File: packages/sdk-go/trace.go
package whyops

import (
	"context"
	"fmt"
	"log"

	"github.com/google/uuid"
)

// Trace is a builder for a single trace (session / conversation).
// Create one via Client.Trace(traceID).
type Trace struct {
	traceID        string
	agentName      string
	apiKey         string
	analyseBaseURL string
	http           *httpClient
	onInit         func(ctx context.Context)
}

func newTrace(traceID, agentName, apiKey, analyseBaseURL string, http *httpClient, onInit func(ctx context.Context)) *Trace {
	return &Trace{
		traceID:        traceID,
		agentName:      agentName,
		apiKey:         apiKey,
		analyseBaseURL: analyseBaseURL,
		http:           http,
		onInit:         onInit,
	}
}

// ─── UserMessage ──────────────────────────────────────────────────────────────

func (t *Trace) UserMessage(ctx context.Context, messages []MessageItem, opts UserMessageOptions) error {
	t.onInit(ctx)
	var meta map[string]any
	if opts.SystemPrompt != "" || len(opts.Tools) > 0 || len(opts.Params) > 0 {
		meta = make(map[string]any)
		if opts.SystemPrompt != "" {
			meta["systemPrompt"] = opts.SystemPrompt
		}
		if len(opts.Tools) > 0 {
			meta["tools"] = opts.Tools
		}
		if len(opts.Params) > 0 {
			meta["params"] = opts.Params
		}
	}
	return t.send(ctx, t.build(EventTypeUserMessage, map[string]any{"messages": messages}, meta, opts.EventOptions))
}

// ─── LLMResponse ─────────────────────────────────────────────────────────────

func (t *Trace) LLMResponse(ctx context.Context, model, provider, content string, opts LLMResponseOptions) error {
	t.onInit(ctx)
	c := map[string]any{"content": content}
	if len(opts.ToolCalls) > 0 {
		c["toolCalls"] = opts.ToolCalls
	}
	if opts.FinishReason != "" {
		c["finishReason"] = opts.FinishReason
	}

	meta := map[string]any{"model": model, "provider": provider}
	if opts.Usage != nil {
		meta["usage"] = opts.Usage
	}
	if opts.LatencyMs > 0 {
		meta["latencyMs"] = opts.LatencyMs
	}
	return t.send(ctx, t.build(EventTypeLLMResponse, c, meta, opts.EventOptions))
}

// ─── LLMThinking ─────────────────────────────────────────────────────────────

type LLMThinkingOptions struct {
	EventOptions
	Signature string
}

func (t *Trace) LLMThinking(ctx context.Context, thinking string, opts LLMThinkingOptions) error {
	t.onInit(ctx)
	c := map[string]any{"type": "thinking", "thinking": thinking}
	if opts.Signature != "" {
		c["signature"] = opts.Signature
	}
	return t.send(ctx, t.build(EventTypeLLMThinking, c, nil, opts.EventOptions))
}

// ─── EmbeddingRequest ────────────────────────────────────────────────────────

func (t *Trace) EmbeddingRequest(ctx context.Context, inputs []string, opts EventOptions) error {
	t.onInit(ctx)
	return t.send(ctx, t.build(EventTypeEmbeddingRequest, map[string]any{"input": inputs}, nil, opts))
}

// ─── EmbeddingResponse ───────────────────────────────────────────────────────

func (t *Trace) EmbeddingResponse(ctx context.Context, model, provider string, embeddingCount, firstDimensions int, opts EmbeddingResponseOptions) error {
	t.onInit(ctx)
	c := map[string]any{
		"object":                   "list",
		"embeddingCount":           embeddingCount,
		"firstEmbeddingDimensions": firstDimensions,
		"encodingFormat":           "float",
	}
	meta := map[string]any{"model": model, "provider": provider}
	if opts.TotalTokens > 0 {
		meta["usage"] = map[string]int{"totalTokens": opts.TotalTokens}
	}
	if opts.LatencyMs > 0 {
		meta["latencyMs"] = opts.LatencyMs
	}
	return t.send(ctx, t.build(EventTypeEmbeddingResponse, c, meta, opts.EventOptions))
}

// ─── ToolCallRequest ─────────────────────────────────────────────────────────

// ToolCallRequest emits a tool_call_request event and returns the spanID.
// Pass the returned spanID to ToolCallResponse to pair them.
func (t *Trace) ToolCallRequest(ctx context.Context, tool string, toolCalls []ToolCallPair, opts ToolCallRequestOptions) (spanID string, err error) {
	t.onInit(ctx)
	if opts.SpanID == "" {
		opts.SpanID = uuid.NewString()
	}
	spanID = opts.SpanID

	c := map[string]any{"toolCalls": toolCalls}
	if opts.RequestedAt != "" {
		c["requestedAt"] = opts.RequestedAt
	}
	meta := map[string]any{"tool": tool}
	if opts.LatencyMs > 0 {
		meta["latencyMs"] = opts.LatencyMs
	}
	err = t.send(ctx, t.build(EventTypeToolCallRequest, c, meta, opts.EventOptions))
	return
}

// ─── ToolCallResponse ────────────────────────────────────────────────────────

func (t *Trace) ToolCallResponse(ctx context.Context, tool, spanID string, toolCalls []ToolCallPair, toolResults map[string]any, opts ToolCallResponseOptions) error {
	t.onInit(ctx)
	opts.SpanID = spanID
	c := map[string]any{"toolCalls": toolCalls, "toolResults": toolResults}
	if opts.RespondedAt != "" {
		c["respondedAt"] = opts.RespondedAt
	}
	meta := map[string]any{"tool": tool}
	if opts.LatencyMs > 0 {
		meta["latencyMs"] = opts.LatencyMs
	}
	return t.send(ctx, t.build(EventTypeToolCallResponse, c, meta, opts.EventOptions))
}

// ─── ToolResult ──────────────────────────────────────────────────────────────

func (t *Trace) ToolResult(ctx context.Context, toolName string, output map[string]any, opts EventOptions) error {
	t.onInit(ctx)
	return t.send(ctx, t.build(EventTypeToolResult, map[string]any{"toolName": toolName, "output": output}, nil, opts))
}

// ─── Error ───────────────────────────────────────────────────────────────────

func (t *Trace) Error(ctx context.Context, message string, opts ErrorOptions) error {
	t.onInit(ctx)
	c := map[string]any{"message": message}
	if opts.Status != 0 {
		c["status"] = opts.Status
	}
	if opts.Stack != "" {
		c["stack"] = opts.Stack
	}
	return t.send(ctx, t.build(EventTypeError, c, nil, opts.EventOptions))
}

// ─── Internal ─────────────────────────────────────────────────────────────────

func (t *Trace) build(eventType EventType, content, metadata any, opts EventOptions) eventPayload {
	var meta map[string]any
	if metadata != nil {
		if m, ok := metadata.(map[string]any); ok {
			meta = m
		}
	}
	return eventPayload{
		EventType:      eventType,
		TraceID:        t.traceID,
		AgentName:      t.agentName,
		SpanID:         opts.SpanID,
		StepID:         opts.StepID,
		ParentStepID:   opts.ParentStepID,
		Timestamp:      opts.Timestamp,
		Content:        content,
		Metadata:       meta,
		IdempotencyKey: opts.IdempotencyKey,
		ExternalUserID: opts.ExternalUserID,
	}
}

func (t *Trace) send(ctx context.Context, payload eventPayload) error {
	url := t.analyseBaseURL + EndpointEventsIngest
	headers := map[string]string{"Authorization": "Bearer " + t.apiKey}

	status, _, err := t.http.post(ctx, url, payload, headers)
	if err != nil {
		log.Printf("%s event send error (%s): %v", LogPrefix, payload.EventType, err)
		return fmt.Errorf("whyops: send: %w", err)
	}
	if status < 200 || status >= 300 {
		log.Printf("%s event send failed: HTTP %d (%s)", LogPrefix, status, payload.EventType)
		return fmt.Errorf("whyops: HTTP %d", status)
	}
	return nil
}
